File: backend/routes/chords.py
from flask import Blueprint, request, jsonify

chords_bp = Blueprint("chords", __name__)


@chords_bp.route("/", methods=["POST"])
def suggest_chords():
    data = request.json
    key = data.get("key", "C").upper()
    mood = data.get("mood", "happy").lower()
    genre = data.get("genre", "general").lower()

    progressions = {
        "general": {
            "happy": {
                "C": [["C", "G", "Am", "F"], ["C", "F", "G", "C"]],
                "C#": [["C#", "G#", "A#m", "F#"], ["C#", "F#", "G#", "C#"]],
                "D": [["D", "A", "Bm", "G"], ["D", "G", "A", "D"]],
                "D#": [["D#", "A#", "Cm", "G#"], ["D#", "G#", "A#", "D#"]],
                "E": [["E", "B", "C#m", "A"], ["E", "A", "B", "E"]],
                "F": [["F", "C", "Dm", "Bb"], ["F", "Bb", "C", "F"]],
                "F#": [["F#", "C#", "D#m", "B"], ["F#", "B", "C#", "F#"]],
                "G": [["G", "D", "Em", "C"], ["G", "C", "D", "G"]],
                "G#": [["G#", "D#", "Fm", "C#"], ["G#", "C#", "D#", "G#"]],
                "A": [["A", "E", "F#m", "D"], ["A", "D", "E", "A"]],
                "A#": [["A#", "F", "Gm", "D#"], ["A#", "D#", "F", "A#"]],
                "B": [["B", "F#", "G#m", "E"], ["B", "E", "F#", "B"]],
                "Bb": [["Bb", "F", "Gm", "Eb"], ["Bb", "Eb", "F", "Bb"]],
                "Eb": [["Eb", "Bb", "Cm", "Ab"], ["Eb", "Ab", "Bb", "Eb"]],
            },
            "sad": {
                "C": [["Am", "F", "C", "G"], ["C", "G", "F", "Am"]],
                "C#": [["A#m", "F#", "C#", "G#"], ["C#", "G#", "F#", "A#m"]],
                "D": [["Bm", "G", "D", "A"], ["D", "A", "G", "Bm"]],
                "D#": [["Cm", "G#", "D#", "A#"], ["D#", "A#", "G#", "Cm"]],
                "E": [["C#m", "A", "E", "B"], ["E", "B", "A", "C#m"]],
                "F": [["Dm", "Bb", "F", "C"], ["F", "C", "Bb", "Dm"]],
                "F#": [["D#m", "B", "F#", "C#"], ["F#", "C#", "B", "D#m"]],
                "G": [["Em", "C", "G", "D"], ["G", "D", "C", "Em"]],
                "G#": [["Fm", "C#", "G#", "D#"], ["G#", "D#", "C#", "Fm"]],
                "A": [["F#m", "D", "A", "E"], ["A", "E", "D", "F#m"]],
                "A#": [["Gm", "D#", "A#", "F"], ["A#", "F", "D#", "Gm"]],
                "B": [["G#m", "E", "B", "F#"], ["B", "F#", "E", "G#m"]],
                "Bb": [["Gm", "Eb", "Bb", "F"], ["Bb", "F", "Eb", "Gm"]],
                "Eb": [["Cm", "Ab", "Eb", "Bb"], ["Eb", "Bb", "Ab", "Cm"]],
            },
            "neutral": {
                "C": [["C", "Am", "F", "G"], ["F", "C", "G", "Am"]],
                "C#": [["C#", "A#m", "F#", "G#"], ["F#", "C#", "G#", "A#m"]],
                "D": [["D", "Bm", "G", "A"], ["G", "D", "A", "Bm"]],
                "D#": [["D#", "Cm", "G#", "A#"], ["G#", "D#", "A#", "Cm"]],
                "E": [["E", "C#m", "A", "B"], ["A", "E", "B", "C#m"]],
                "F": [["F", "Dm", "Bb", "C"], ["Bb", "F", "C", "Dm"]],
                "F#": [["F#", "D#m", "B", "C#"], ["B", "F#", "C#", "D#m"]],
                "G": [["G", "Em", "C", "D"], ["C", "G", "D", "Em"]],
                "G#": [["G#", "Fm", "C#", "D#"], ["C#", "G#", "D#", "Fm"]],
                "A": [["A", "F#m", "D", "E"], ["D", "A", "E", "F#m"]],
                "A#": [["A#", "Gm", "D#", "F"], ["D#", "A#", "F", "Gm"]],
                "B": [["B", "G#m", "E", "F#"], ["E", "B", "F#", "G#m"]],
                "Bb": [["Bb", "Gm", "Eb", "F"], ["Eb", "Bb", "F", "Gm"]],
                "Eb": [["Eb", "Cm", "Ab", "Bb"], ["Ab", "Eb", "Bb", "Cm"]],
            },
            "angry": {
                "C": [["C5", "G5", "F5", "E5"], ["C5", "F5", "G5", "C5"]],
                "G": [["G5", "D5", "C5", "B5"], ["G5", "C5", "D5", "G5"]],
            },
            "romantic": {
                "C": [["Cmaj7", "Am7", "Dm7", "G7"], ["Fmaj7", "Em7", "Am7", "Dm7"]],
                "G": [["Gmaj7", "Em7", "Am7", "D7"], ["Cmaj7", "Bm7", "Em7", "Am7"]],
            },
            "excited": {
                "C": [["C", "E7", "F", "G"], ["C", "G", "Am", "F"]],
                "G": [["G", "B7", "C", "D"], ["G", "D", "Em", "C"]],
            },
            "peaceful": {
                "C": [["Cmaj7", "Fmaj7", "Am7", "G7"], ["Fmaj7", "Cmaj7", "G7", "Am7"]],
                "G": [["Gmaj7", "Cmaj7", "Em7", "D7"], ["Cmaj7", "Gmaj7", "D7", "Em7"]],
            },
            "melancholic": {
                "C": [["Am", "F", "C", "G"], ["F", "C", "G", "Am"]],
                "G": [["Em", "C", "G", "D"], ["C", "G", "D", "Em"]],
            },
            "nostalgic": {
                "C": [["C", "Am", "F", "G"], ["Am", "F", "C", "G"]],
                "G": [["G", "Em", "C", "D"], ["Em", "C", "G", "D"]],
            },
            "dark": {
                "C": [["Cm", "Ab", "Bb", "Gm"], ["Gm", "Bb", "Ab", "Cm"]],
                "G": [["Gm", "Eb", "F", "Cm"], ["Cm", "F", "Eb", "Gm"]],
            },
        },
        "pop": {
            "happy": {
                "C": [["C", "Am", "F", "G"], ["F", "G", "C", "Am"]],
                "G": [["G", "Em", "C", "D"], ["C", "D", "G", "Em"]],
            },
            "sad": {
                "C": [["Am", "F", "C", "G"], ["F", "C", "G", "Am"]],
                "G": [["Em", "C", "G", "D"], ["C", "G", "D", "Em"]],
            },
        },
        "rock": {
            "happy": {
                "C": [["C5", "G5", "F5", "G5"], ["C5", "F5", "G5", "C5"]],
                "G": [["G5", "D5", "C5", "D5"], ["G5", "C5", "D5", "G5"]],
            },
            "angry": {
                "C": [["C5", "G5", "F5", "E5"], ["C5", "F5", "G5", "C5"]],
                "G": [["G5", "D5", "C5", "B5"], ["G5", "C5", "D5", "G5"]],
            },
        },
        "jazz": {
            "romantic": {
                "C": [["Cmaj7", "Am7", "Dm7", "G7"], ["Fmaj7", "Em7", "Am7", "Dm7"]],
                "G": [["Gmaj7", "Em7", "Am7", "D7"], ["Cmaj7", "Bm7", "Em7", "Am7"]],
            },
            "peaceful": {
                "C": [["Cmaj7", "Fmaj7", "Am7", "G7"], ["Fmaj7", "Cmaj7", "G7", "Am7"]],
                "G": [["Gmaj7", "Cmaj7", "Em7", "D7"], ["Cmaj7", "Gmaj7", "D7", "Em7"]],
            },
        },
        "blues": {
            "excited": {
                "C": [["C7", "F7", "G7", "C7"], ["C7", "G7", "F7", "C7"]],
                "G": [["G7", "C7", "D7", "G7"], ["G7", "D7", "C7", "G7"]],
            },
        },
        "hiphop": {
            "happy": {
                "C": [["C7", "F7", "G7", "C7"]],
                "G": [["G7", "C7", "D7", "G7"]],
            },
            "sad": {
                "C": [["Am7", "Dm7", "G7", "Cmaj7"]],
                "G": [["Em7", "Am7", "D7", "Gmaj7"]],
            },
            "neutral": {
                "C": [["Cmaj7", "Am7", "Dm7", "G7"]],
                "G": [["Gmaj7", "Em7", "Am7", "D7"]],
            },
        },
        "electronic": {
            "happy": {
                "C": [["C", "G", "Am", "F"]],
                "G": [["G", "D", "Em", "C"]],
            },
            "sad": {
                "C": [["Am", "F", "C", "G"]],
                "G": [["Em", "C", "G", "D"]],
            },
        },
        "metal": {
            "angry": {
                "C": [["C5", "G5", "F5", "E5"]],
                "G": [["G5", "D5", "C5", "B5"]],
            },
        },
        "folk": {
            "happy": {
                "C": [["C", "F", "G", "C"]],
                "G": [["G", "C", "D", "G"]],
            },
        },
        "country": {
            "happy": {
                "C": [["C", "G", "F", "C"]],
                "G": [["G", "D", "C", "G"]],
            },
        },
    }

    genre_progressions = progressions.get(genre, progressions.get("general", {}))
    mood_progressions = genre_progressions.get(mood, {})

    chords = mood_progressions.get(key)

    if not chords:
        if mood_progressions:
            chords = next(iter(mood_progressions.values()))
        else:
            chords = [["C", "G", "Am", "F"]]

    return jsonify({"chords": chords})

File: frontend/src/components/Chords.js
import React, {useState, useEffect} from "react";
import axios from "axios";
import {File, Track} from "jsmidgen";
import "./Chords.css";

const TICKS_PER_BEAT = 480; // Define constant here
const NOTE_NAMES = ['C', 'C#', 'D', 'D#', 'E', 'F', 'F#', 'G', 'G#', 'A', 'A#', 'B'];
const CHORD_INTERVALS = {
    major: [0, 4, 7],
    minor: [0, 3, 7],
    major7: [0, 4, 7, 11],
    minor7: [0, 3, 7, 10],
    dominant7: [0, 4, 7, 10],
    diminished: [0, 3, 6],
    augmented: [0, 4, 8],
    sus2: [0, 2, 7],
    sus4: [0, 5, 7],
};

function parseChordName(chordName) {
    const regex = /^([A-G]#?b?)(.*)$/;
    const match = chordName.match(regex);
    if (!match) return null;
    const root = match[1];
    let quality = match[2].toLowerCase();
    if (quality === "") quality = "major";
    else if (quality === "m") quality = "minor";
    else if (quality === "maj7") quality = "major7";
    else if (quality === "7") quality = "dominant7";

    return {root, quality};
}

function generateChordNotes(root, quality = "major", octave = 4) {
    const rootIndex = NOTE_NAMES.indexOf(root);
    if (rootIndex === -1) return [];
    const intervals = CHORD_INTERVALS[quality];
    if (!intervals) return [];

    return intervals.map(interval => {
        const noteIndex = (rootIndex + interval) % 12;
        const noteOctave = octave + Math.floor((rootIndex + interval) / 12);
        return NOTE_NAMES[noteIndex] + noteOctave;
    });
}

function invertChord(notes, inversion = 0) {
    const inverted = [...notes];
    for (let i = 0; i < inversion; i++) {
        const note = inverted.shift();
        const pitch = note.slice(0, -1);
        const octave = parseInt(note.slice(-1), 10) + 1;
        inverted.push(pitch + octave);
    }
    return inverted;
}

const stringToUint8Array = (str) => {
    const arr = new Uint8Array(str.length);
    for (let i = 0; i < str.length; i++) {
        arr[i] = str.charCodeAt(i);
    }
    return arr;
};
const Chords = () => {
    const [key, setKey] = useState("C");
    const [mood, setMood] = useState("happy");
    const [genre, setGenre] = useState("general");
    const [chordProgression, setChordProgression] = useState([]);
    const [loading, setLoading] = useState(false);
    const [chordError, setChordError] = useState("");
    const [fallbackUsed, setFallbackUsed] = useState(false);
    const [fallbackInfo, setFallbackInfo] = useState("");
    const [importedMidiName, setImportedMidiName] = useState("");
    const [audioFileName, setAudioFileName] = useState("");
    const melodySequence = ["E4", "D4", "F4", "G4"];
    const generateChords = async (keyParam = key, moodParam = mood, genreParam = genre) => {
        setLoading(true);
        setChordError("");
        setFallbackUsed(false);
        setFallbackInfo("");
        try {
            const response = await axios.post("/api/chords/", {key: keyParam, mood: moodParam, genre: genreParam});
            if (response.data && Array.isArray(response.data.chords) && response.data.chords.length > 0) {
                setChordProgression(response.data.chords[0]);
                const requestedKey = keyParam.toUpperCase();
                const firstChordRoot = parseChordName(response.data.chords[0][0])?.root.toUpperCase();
                if (firstChordRoot && firstChordRoot !== requestedKey) {
                    setFallbackUsed(true);
                    setFallbackInfo(`No progression found for key "${requestedKey}". Showing progression for key "${firstChordRoot}" instead.`);
                }
            } else {
                setChordError("No chord progression returned from the API.");
                setChordProgression([]);
            }
        } catch (error) {
            setChordError(error.response?.data?.error || "Failed to generate chords. Please try again.");
            setChordProgression([]);
        } finally {
            setLoading(false);
        }
    };

    useEffect(() => {
        generateChords(key, mood, genre);
    }, [key, mood, genre]);

    const downloadMidi = () => {
        if (!chordProgression || chordProgression.length === 0) return;

        const BEATS_PER_BAR = 4;    // assuming 4/4 time
        const TICKS_PER_BAR = TICKS_PER_BEAT * BEATS_PER_BAR;

        const file = new File();
        file.header.setTicksPerBeat(TICKS_PER_BEAT);

        const chordTrack = new Track();
        const bassTrack = new Track();
        const melodyTrack = new Track();
        file.addTrack(chordTrack);
        file.addTrack(bassTrack);
        file.addTrack(melodyTrack);

        const bars = 8; // fixed length
        const chordsPerBar = 1; // one chord per bar
        const totalChords = bars * chordsPerBar;

        const velocity = 100;

        // Rhythmic pattern for melody (in ticks)
        const melodyRhythm = [TICKS_PER_BEAT, TICKS_PER_BEAT / 2, TICKS_PER_BEAT / 2, TICKS_PER_BEAT];

        // Scale notes for passing tones (C major scale example)
        const scaleNotes = ["C4", "D4", "E4", "F4", "G4", "A4", "B4", "C5"];

        for (let i = 0; i < totalChords; i++) {
            const chordName = chordProgression[i % chordProgression.length];
            const baseDelay = i * TICKS_PER_BAR;

            const parsed = parseChordName(chordName);
            if (!parsed) {
                console.warn(`Skipping invalid chord name: ${chordName}`);
                continue;
            }

            let chordNotes = generateChordNotes(parsed.root, parsed.quality, 4);
            chordNotes = invertChord(chordNotes, i % chordNotes.length);

            // Arpeggiated chord notes with rhythmic variation
            let tickCursor = baseDelay;
            chordNotes.forEach((note, idx) => {
                const duration = melodyRhythm[idx % melodyRhythm.length];
                chordTrack.addNote(0, note, duration, tickCursor, velocity);
                tickCursor += duration;
            });

            // Bass note on root, whole bar duration
            bassTrack.addNote(0, parsed.root + "2", TICKS_PER_BAR, baseDelay, velocity);

            // Melody motif using chord tones and passing scale notes
            let melodyTick = baseDelay + TICKS_PER_BEAT / 2;
            for (let j = 0; j < 4; j++) {
                const isChordTone = j % 2 === 0;
                let note;
                if (isChordTone) {
                    note = chordNotes[j % chordNotes.length];
                } else {
                    const rootIndex = scaleNotes.indexOf(parsed.root + "4");
                    if (rootIndex === -1) {
                        note = scaleNotes[0];
                    } else {
                        note = scaleNotes[rootIndex + (j % 2 === 0 ? 1 : -1)] || scaleNotes[rootIndex];
                    }
                }
                const duration = melodyRhythm[j % melodyRhythm.length];
                melodyTrack.addNote(0, note, duration, melodyTick, velocity - 20);
                melodyTick += duration;
            }
        }

        const midiData = file.toBytes();
        const midiArray = stringToUint8Array(midiData);
        const blob = new Blob([midiArray], {type: "audio/midi"});
        const url = URL.createObjectURL(blob);
        const a = document.createElement("a");
        a.href = url;
        a.download = "melodic_chord_progression_8bars.mid";
        a.click();
        URL.revokeObjectURL(url);
    };

    const handleMidiImport = (e) => {
        const file = e.target.files[0];
        if (file) {
            setImportedMidiName(file.name);
            // TODO: Implement MIDI parsing if needed
        }
    };

    const handleAudioUpload = (e) => {
        const file = e.target.files[0];
        if (file) {
            setAudioFileName(file.name);
            // TODO: Implement audio processing if needed
        }
    };

    return (
        <div className="midi-integration">
            <h2>MIDI & Audio Integration</h2>

            <div className="export-section">
                <h3>Export Chord Progression with Melody as MIDI</h3>

                <div style={{marginBottom: "10px"}}>
                    <label htmlFor="key-select">Select Key: </label>
                    <select id="key-select" value={key} onChange={e => setKey(e.target.value)} disabled={loading}>
                        <option value="C">C</option>
                        <option value="C#">C# / Db</option>
                        <option value="D">D</option>
                        <option value="D#">D# / Eb</option>
                        <option value="E">E</option>
                        <option value="F">F</option>
                        <option value="F#">F# / Gb</option>
                        <option value="G">G</option>
                        <option value="G#">G# / Ab</option>
                        <option value="A">A</option>
                        <option value="A#">A# / Bb</option>
                        <option value="B">B</option>
                        <option value="Bb">Bb</option>
                        <option value="Eb">Eb</option>
                    </select>

                    <label htmlFor="mood-select" style={{marginLeft: "20px"}}>Select Mood: </label>
                    <select id="mood-select" value={mood} onChange={e => setMood(e.target.value)} disabled={loading}>
                        <option value="happy">Happy</option>
                        <option value="sad">Sad</option>
                        <option value="neutral">Neutral</option>
                        <option value="angry">Angry</option>
                        <option value="romantic">Romantic</option>
                        <option value="excited">Excited</option>
                        <option value="peaceful">Peaceful</option>
                        <option value="melancholic">Melancholic</option>
                        <option value="nostalgic">Nostalgic</option>
                        <option value="dark">Dark</option>
                    </select>

                    <label htmlFor="genre-select" style={{marginLeft: "20px"}}>Select Genre: </label>
                    <select id="genre-select" value={genre} onChange={e => setGenre(e.target.value)} disabled={loading}>
                        <option value="general">General</option>
                        <option value="pop">Pop</option>
                        <option value="rock">Rock</option>
                        <option value="jazz">Jazz</option>
                        <option value="blues">Blues</option>
                        <option value="country">Country</option>
                        <option value="folk">Folk</option>
                        <option value="metal">Metal</option>
                        <option value="hiphop">Hip Hop</option>
                        <option value="electronic">Electronic</option>
                    </select>

                    <button
                        className="action-button"
                        onClick={() => generateChords(key, mood, genre)}
                        disabled={loading}
                        style={{marginLeft: "20px"}}
                    >
                        {loading ? "Generating..." : "Generate Chords"}
                    </button>
                </div>

                {fallbackUsed && (
                    <div className="fallback-warning" style={{
                        marginTop: "10px",
                        color: "#b35c00",
                        backgroundColor: "#fff4e5",
                        padding: "10px",
                        borderRadius: "6px"
                    }}>
                        {fallbackInfo}
                    </div>
                )}

                <h4>Current Chord Progression:</h4>
                {chordProgression.length > 0 ? (
                    <div className="chord-list">
                        {chordProgression.map((chord, idx) => (
                            <span key={idx} className="chord-badge">{chord}</span>
                        ))}
                    </div>
                ) : (
                    <p>No chords available</p>
                )}

                <button
                    className="action-button"
                    onClick={downloadMidi}
                    disabled={chordProgression.length === 0}
                    style={{marginTop: "10px"}}
                >
                    Download as MIDI
                </button>
            </div>

            <div className="notes-section" style={{
                marginTop: "30px",
                padding: "15px",
                backgroundColor: "#fff4f2",
                borderRadius: "8px",
                color: "#333"
            }}>
                <h3>Helpful Notes for Generating Chords</h3>
                <ul>
                    <li><strong>Keys:</strong> Choose a key that fits the vocal range or mood of your song. Common keys
                        like C, G, and D are easier to play on guitar and piano.
                    </li>
                    <li><strong>Moods:</strong> Select a mood that matches the emotional tone you want. For example,
                        "happy" keys often use major chords, while "sad" moods lean towards minor chords.
                    </li>
                    <li><strong>Genres:</strong> Different genres have characteristic chord progressions. For example,
                        blues often uses dominant 7th chords, jazz uses complex chords like major7 and minor7, and pop
                        tends to use simple major/minor progressions.
                    </li>
                    <li><strong>Experiment:</strong> Try different combinations to find unique progressions that inspire
                        your creativity.
                    </li>
                    <li><strong>Limitations:</strong> The backend currently supports a limited set of keys and moods. If
                        you select unsupported options, you may get no progressions.
                    </li>
                </ul>
            </div>

            <div className="import-section" style={{marginTop: "30px"}}>
                <h3>Import a MIDI File</h3>
                <input type="file" accept=".mid,.midi" onChange={handleMidiImport}/>
                {importedMidiName && <p>Imported MIDI: {importedMidiName}</p>}
            </div>

            <div className="audio-section" style={{marginTop: "30px"}}>
                <h3>Upload Audio Stem</h3>
                <input type="file" accept="audio/*" onChange={handleAudioUpload}/>
                {audioFileName && <p>Uploaded Audio: {audioFileName}</p>}
            </div>
        </div>
    );
};
export default Chords;